node_monitor/node_monitor.py

```diff
@@ -173,6 +175,35 @@ def broadcast_status_report(self) -> None:
                     if err2 is not None:
                         logging.error(f"TelegramBot.send_message() failed with error: {err2}")
             # - - - - - - - - - - - - - - - - -
+    
+
+    def update_node_provider_lookup_if_new(
+            self, 
+            override_api_data: ic_api.NodeProviders | None = None) -> None:
+        """Fetches the current node providers from the ic-api and compares
+        them to what is currently in the node_provider_lookup table in the 
+        database. If there is a new node provider in the API, they will be
+        added to our database.
+
+        Args:
+            override_data: If provided, this arg will be used instead of 
+                live fetching Node Providers from the ic-api. Useful for testing.
+        """
+        data = override_api_data if override_api_data else ic_api.get_node_providers()
+        node_providers_in_database = self.node_provider_db.get_node_providers_as_dict()
+
+        ic_api_principals = set(node_provider.principal_id 
+                                for node_provider in data.node_providers)
+        database_principals = set(node_providers_in_database.keys())
```

Comment:
3. You're dealing with two different types and this makes it hard to read.
I recommend to create the variable node_providers_api as a list of dictionaries from the contents of data.node_providers. That way you don't have to do the weird set comprehension